src/robot/parsing/model/visitor.py

```diff
@@ -14,27 +14,49 @@
 #  limitations under the License.
 
 import ast
+from typing import Any, Callable, Dict, Optional, Type, Union
 
 from .statements import Node
 
 
+class _NotSet:
+    pass
+
+
 class VisitorFinder:
+    __NOT_SET = _NotSet()
+    __cls_cache: Dict[Type[Any], Union[Callable[..., Any], None, _NotSet]]
 
-    def _find_visitor(self, cls):
-        if cls is ast.AST:
+    def __new__(cls, *_args: Any, **_kwargs: Any):  # type: ignore[no-untyped-def]
+        # create cache on class level to avoid creating it for each instance
+        cls.__cls_cache = {}
+        return super().__new__(cls)
+
+    @classmethod
+    def __find_visitor(cls, node_cls: Type[Any]) -> Optional[Callable[..., Any]]:
+        if node_cls is ast.AST:
             return None
-        method = 'visit_' + cls.__name__
-        if hasattr(self, method):
-            return getattr(self, method)
-        # Forward-compatibility.
-        if method == 'visit_Return' and hasattr(self, 'visit_ReturnSetting'):
-            return getattr(self, 'visit_ReturnSetting')
-        for base in cls.__bases__:
-            visitor = self._find_visitor(base)
-            if visitor:
-                return visitor
+        method_name = "visit_" + node_cls.__name__
+        method = getattr(cls, method_name, None)
+        if callable(method):
+            return method  # type: ignore[no-any-return]
+        if method_name == "visit_Return":
+            method = getattr(cls, "visit_ReturnSetting", None)
+            if callable(method):
+                return method  # type: ignore[no-any-return]
+        for base in node_cls.__bases__:
+            method = cls._find_visitor(base)
+            if method:
+                return method
         return None
 
+    @classmethod
+    def _find_visitor(cls, node_cls: Type[Any]) -> Optional[Callable[..., Any]]:
+        result = cls.__cls_cache.get(node_cls, cls.__NOT_SET)
+        if result is cls.__NOT_SET:
```

Comment:
no, this is `node_cls in cls.__cls_cache' and then `cls.__cls_cache[node_cls]` in one call this saves time because an `is` check is faster then a dictionary operation. 

NOT_SET is needed because if we did not found a method then we get None